### DISCUSSION TITLE
Feature Request:  GUI Editable File Provisioned Alerts in 9.1.x

### DISCUSSION THREAD

**Discussion by sarasensible** [grafana/grafana#53490] [Alerting]
Hello!  I am very excited to see the file provisioning for Alert Rules being added in 9.1.x beginning with this PR https://github.com/grafana/grafana/pull/51635 .  I went ahead and tried this out (configuration below) and while it works great I see that the end resulting alert rules are not editable from the UI.  I consider the provisioned rules a starting point for our Grafana users who definitely will have tweaks they want to make from time to time, which can then be exported back to code.  This is how it works with editable dashboards - there are provisioned dashboards, but they can be edited via the UI and then re-exported back for IAC.
Provisioned Rules
<img width="1360" alt="Screen Shot 2022-08-09 at 4 30 20 PM" src="https://user-images.githubusercontent.com/97973481/183754967-21f78818-2313-4dec-8a40-eb40163ab6de.png">
View only
<img width="933" alt="Screen Shot 2022-08-09 at 4 48 05 PM" src="https://user-images.githubusercontent.com/97973481/183758007-082ffe69-e9af-4aff-998a-7946df0b1dcd.png">

This would be a killer feature and would support our adoption of Grafana alerting.

Edit:  This [issue](https://github.com/grafana/grafana/issues/45297) looks related and implies that at least export/import will be supported, so that implies editing as well hopefully.  According to this [issue](https://github.com/grafana/grafana/issues/40647) it seems some users have the opposite requirement that provisioned alerts be blocked from being editable.  My hope is that this could become a feature toggle as is the case with dashboards.

Here is how I provisioned the alert rules using Terraform and Helm via the functionality available in the beta release.

main.tf
```
resource "kubernetes_config_map" "grafana-alerts" {
  metadata {
    name = "grafana-alerts"
    namespace = "monitoring"
  }

  data = {
    "grafana-alerts.json" = file("${path.module}/helm/prometheus/grafana-alerts.json")
  }
}

resource "helm_release" "monitoring" {
  depends_on = [kubernetes_config_map.grafana-alerts]
  name  = "prometheus-stack-1"
  repository = "https://prometheus-community.github.io/helm-charts"
  chart = "kube-prometheus-stack"
  namespace = "monitoring"
  values  = [templatefile("${path.module}/helm/prometheus/prometheus.tftpl", {})]
}
```

prometheus.tftpl
```
grafana:
  image:
    tag: 9.1.0-beta1
  extraConfigmapMounts:
  - name: grafana-alerts
    mountPath: /etc/grafana/provisioning/alerting/grafana-alerts.json
    subPath: grafana-alerts.json
    configMap: grafana-alerts
    readOnly: true
```

The JSON for the alerts is almost directly translatable from the API output with a few minor tweaks (which might depend on the version - I retrieved the rules from 9.0.3):
`curl -X GET -H "Authorization: Bearer $GRAFANA_TOKEN" -H "Content-type: application/json" "http://localhost:3000/api/ruler/grafana/api/v1/rules/" | jq > grafana-alerts.json`

As an aside, I attempted to use the new rules provisioning API endpoint available at https://grafana.com/docs/grafana/latest/developers/http_api/alerting_provisioning/ as recommended by @JohnnyQQQQ in https://github.com/grafana/grafana/issues/43066#issuecomment-1185368660 , however I was not able to figure out the permissions schema for this and kept running into errors such as `{"accessErrorId":"ACE0032154373","message":"You'll need additional permissions to perform this action. Permissions needed: alert.provisioning:read","title":"Access denied"}`

Update on this:  I found out the issue with the new API was the dashboard folder referred to as `Folder-name` needed to be created prior to running this Terraform.  I actually updated my workaround Terraform to use basic auth instead of a token (since a brand new Grafana doesn't have the API key created) and added folder creation.

Output Structure
```
{
  "Folder-name": [
    {
      "name": "Alert-group name",
      "interval": "1m",
      "rules": [
        {
          "grafana_alert": {
             "id": 6,
            "orgId": 1,
            "title": "alert-title",
            "condition": "B",
            "data": [
```
Tweaks:
1. Change the top level Folder-name to "groups"
2. Add `"apiVersion": 1,` to the top
3. Add the "folder" key with the new folder name under the rule bodies
4. Remove the top level "grafana_alert" object from each rule
5. Ensure all the "relativeTimeRange" objects have a "from" and "to" that are different, i.e.
```
"relativeTimeRange": {
    "from": 0,
    "to": 0
  }
```
should be
```
"relativeTimeRange": {
    "from": 300,
    "to": 0
  }
```
( or something similar )
6.  Every alert output by the API has a UID, leave this alone - do not remove it - for the alert to provision properly

grafana-alerts.json
```
{
	"apiVersion": 1,
	"groups": [{
		"name": "Engine - Error Rate",
		"folder": "General",
		"interval": "1m",
		"rules": [{
			"expr": "",
			"for": "5m",
			"labels": {
				"alertname": "Engine - Error Rate",
				"severity": "high"
			},
			"annotations": {
				"__alertId__": "4",
				"__dashboardUID__": "zROzbST7z",
				"__panelId__": "2",
				"description": "Elevated Error Rates Observed in \"engine\"",
				"message": "Elevated Error Rates Observed in \"engine\"",
				"summary": "Elevated Error Rates Observed in \"engine\""
			},
			"title": "Engine - Error Rate",
			"condition": "B",
			"data": [{
					"refId": "A",
					"queryType": "",
					"relativeTimeRange": {
						"from": 300,
						"to": 0
					},
					"datasourceUid": "Prometheus",
					"model": {
						"expr": "rate(http_requests_total{job=\"engine\", status!=\"2xx\"}[1m]) / 60",
						"intervalMs": 1000,
						"legendFormat": "{{ method }} :: {{ handler }} - {{ pod }}",
						"maxDataPoints": 43200,
						"refId": "A"
					}
				},
				{
					"refId": "B",
					"queryType": "",
					"relativeTimeRange": {
						"from": 300,
						"to": 0
					},
					"datasourceUid": "-100",
					"model": {
						"conditions": [{
							"evaluator": {
								"params": [
									5
								],
								"type": "gt"
							},
							"operator": {
								"type": "and"
							},
							"query": {
								"params": [
									"A"
								]
							},
							"reducer": {
								"type": "avg"
							}
						}],
						"intervalMs": 1000,
						"maxDataPoints": 43200,
						"refId": "B",
						"type": "classic_conditions"
					}
				}
			],
			"intervalSeconds": 60,
			"uid": "zPPjDmiVk",
			"version": 2,
			"namespace_uid": "OceOdmi4k",
			"namespace_id": 135,
			"rule_group": "Engine - Error Rate",
			"no_data_state": "NoData",
			"exec_err_state": "Alerting"
		}]
	}]
}
```

Thanks so much in advance!

**Comment by sarasensible** [comment]
I created a workaround in Terraform that uses the API to create editable alerts.  It only creates them if no alert rules are present, and if all rules are deleted it will recreate them.  It will not update them if they are edited independently.  This script only works for one top level folder.  The JSON content should be created under ./helm/prometheus/grafana-alerts.json and should be the content from
`curl -X GET -H "Authorization: Bearer $GRAFANA_TOKEN" -H "Content-type: application/json" "http://${host}/api/ruler/grafana/api/v1/rules/" | jq > grafana-alerts.json`
when rules are created manually.  The folder name passed to the Terraform template needs to match the folder name from the grafana-alerts.json file in order for this to work (in this case I used "Custom" - "General" does not work as this is a [special reserved folder](https://grafana.com/docs/grafana/latest/developers/http_api/folder/#a-note-about-the-general-folder)).

Bash template script inspired by https://community.grafana.com/t/ngalert-grafana-8-alert-feature-how-to-export-import-alerts-as-yml-json/51677/22

main.tf
```
resource "helm_release" "monitoring" {
  depends_on = [kubernetes_config_map.grafana-alerts]
  name  = "prometheus-stack-1"
  repository = "https://prometheus-community.github.io/helm-charts"
  chart = "kube-prometheus-stack"
  namespace = "monitoring"
  values  = [templatefile("${path.module}/helm/prometheus/prometheus.tftpl", {})]
}

data "http" "rules" {
  depends_on = [helm_release.monitoring]
  url = "https://${var.username}:${var.password}@${local.grafana_domain}/api/ruler/grafana/api/v1/rules/"

  request_headers = {
    Content-type = "application/json"
    Accept = "application/json"
  }
}

data "template_file" "upload-grafana-alerts" {
  template = file("${path.module}/helm/prometheus/upload-grafana-alerts.tftpl")

  vars = {
    folder_name = "Custom"
    host = local.grafana_domain
    method = "POST"
    grafana_path = "${path.module}/helm/prometheus/grafana-alerts.json"
    username = var.username
    password = var.password
  }
}

resource "null_resource" "execute-create-grafana-alerts" {
  count = data.http.rules.body == "{}" ? 1 : 0
  depends_on = [helm_release.monitoring]
  triggers = {
    script_change = data.template_file.upload-grafana-alerts.rendered
  }
  provisioner "local-exec" {
    command = data.template_file.upload-grafana-alerts.rendered
  }
}
```

upload-grafana-alerts.tftpl
```
export ALERTS_JSON_PATH=${grafana_path}; ALERT_OBJECT=$(jq -c '.${folder_name}[0] | del(.rules[].grafana_alert.uid)' $ALERTS_JSON_PATH); ALERT_NAME=$(jq -c '.${folder_name}[0].name' $ALERTS_JSON_PATH); echo "Creating $ALERT_NAME...\n"; curl -X ${method} -H "Content-type: application/json" -H "Accept: application/json" 'https://${username}:${password}@${host}/api/ruler/grafana/api/v1/rules/${folder_name}' -d "$ALERT_OBJECT"; echo "\n";
```

Update:  This Terraform only works if the "Custom" folder is created prior to running.  For a fully automated solution you can use the following as well, with a `depends_on = [null_resource.execute-create-grafana-folder]` added to the alert rules TF:

main.tf
```
data "http" "folders" {
  url = "https://${var.username}:${var.password}@${local.grafana_domain}/api/folders"

  request_headers = {
    Content-type = "application/json"
    Accept = "application/json"
  }
}

data "template_file" "upload-grafana-folders" {
  template = file("${path.module}/helm/prometheus/upload-grafana-resource.tftpl")

  vars = {
    host = "${local.grafana_domain}"
    method = "POST"
    post_body = file("${path.module}/helm/prometheus/grafana-folders.json")
    username = var.username
    password = var.password
    path = "folders"
  }
}

resource "null_resource" "execute-create-grafana-folder" {
  count = var.env == "prod" && data.http.folders.body == "[]" ? 1 : 0
  triggers = {
    always_exec = timestamp()
  }
  provisioner "local-exec" {
    command = data.template_file.upload-grafana-folders.rendered
  }
}
```

upload-grafana-resource.tftpl
```
curl -X ${method} -H "Content-type: application/json" -H "Accept: application/json" 'https://${username}:${password}@${host}/api/${path}' -d '${post_body}'; echo "\n";

```

grafana-folders.json
```
{"title":"Custom","canSave":true,"canEdit":true,"canAdmin":true}
```

**Comment by sarasensible** [comment]
I additionally got provisioning to work for contact points and notification policies using this method.

main.tf
```
data "http" "contacts" {
  url = "https://${local.grafana_domain}/api/alertmanager/grafana/config/api/v1/alerts"

  request_headers = {
    Authorization = "Bearer ${data.aws_ssm_parameter.grafana-token.value}"
    Content-type = "application/json"
    Accept = "application/json"
  }
}

data "aws_ssm_parameter" "pd-integration-key" {
  name = "/cluster/${var.env}/PD_INTEGRATION_KEYS"
}

data "template_file" "grafana-contact-points" {
  template = file("${path.module}/helm/prometheus/grafana-contact-points.tftpl")

  vars = {
    pd_integration_backend-daytime = jsondecode(data.aws_ssm_parameter.pd-integration-key.value)["backend_daytime"]
    pd_integration_backend-critical = jsondecode(data.aws_ssm_parameter.pd-integration-key.value)["backend_critical"]
  }
}

data "template_file" "upload-grafana-contact-points" {
  template = file("${path.module}/helm/prometheus/upload-grafana-contacts.tftpl")

  vars = {
    host = local.grafana_domain
    method = "POST"
    post_body = data.template_file.grafana-contact-points.rendered
    token = data.aws_ssm_parameter.grafana-token.value
    path = "alertmanager/grafana/config/api/v1/alerts"
  }
}

## The count parameter here assumes you start off with the default provisioned contact point of grafana email
resource "null_resource" "execute-create-grafana-contacts" {
  count = length(jsondecode(data.http.contacts.body)["alertmanager_config"]["receivers"])==1 ? 1 : 0
  triggers = {
    contacts_change = data.template_file.grafana-contact-points.rendered
  }
  provisioner "local-exec" {
    command = data.template_file.upload-grafana-contact-points.rendered
  }
}
```

grafana-contact-points.tftpl
```
{
  "template_files": {},
  "alertmanager_config": {
    "route": {
      "receiver": "grafana-default-email",
      "routes": [
        {
          "receiver": "Backend-Critical",
          "object_matchers": [
            [
              "severity",
              "=",
              "high"
            ]
          ],
          "continue": false,
          "group_by": [],
          "routes": [],
          "mute_time_intervals": []
        },
        {
          "continue": false,
          "group_by": [],
          "object_matchers": [
            [
              "severity",
              "=",
              "warning"
            ]
          ],
          "routes": [],
          "mute_time_intervals": [],
          "receiver": "Backend-Daytime"
        }
      ],
      "continue": false,
      "group_by": [],
      "mute_time_intervals": []
    },
    "templates": null,
    "receivers": [
      {
        "name": "grafana-default-email",
        "grafana_managed_receiver_configs": [
          {
            "uid": "P9PYm5i4k",
            "name": "email receiver",
            "type": "email",
            "disableResolveMessage": false,
            "settings": {
              "addresses": "<example@email.com>"
            },
            "secureFields": {}
          }
        ]
      },
      {
        "name": "Backend-Daytime",
        "grafana_managed_receiver_configs": [
          {
            "settings": {},
            "secureSettings": {
              "integrationKey": "${pd_integration_backend-daytime}"
            },
            "type": "pagerduty",
            "name": "Backend-Daytime",
            "disableResolveMessage": false
          }
        ]
      },
      {
        "name": "Backend-Critical",
        "grafana_managed_receiver_configs": [
          {
            "settings": {},
            "secureSettings": {
              "integrationKey": "${pd_integration_backend-critical}"
            },
            "type": "pagerduty",
            "name": "Backend-Critical",
            "disableResolveMessage": false
          }
        ]
      }
    ]
  }
}
```

upload-grafana-contacts.tftpl
```
curl -X ${method} -H "Authorization: Bearer ${token}" -H "Content-type: application/json" -H "Accept: application/json" 'https://${host}/api/${path}' -d '${post_body}'; echo "\n";
```

**Comment by JohnnyQQQQ** [comment]
Thanks for opening this Feature Request. 

When doing changes in the UI, on each restart the state from the disk would be replayed -- if the new changes made in the UI are not there yet, the user will lose all their changes. We want predictable behavior and a single source of truth for the data to prevent those nasty things from happening. So currently, we don't plan to change this, as in the past, this has caused a lot of trouble.

I can see that you are using a lot of TF. We will update our provider with the new alerting objects soon, perhaps this helps a bit.

**Reply by alexweav** [reply]
To second this, dashboards have really wanted to move _away_ from making provisioned objects UI-editable for quite some time now. This is one of our most common foot-guns for users, they make changes in a dashboard and then have those changes be mysteriously missing at some unknown point in the future. This ends up happening because either the files got re-provisioned due to a restart, or someone decided to re-run the Terraform apply which brought everything back to the state in TF.

**Reply by sarasensible** [reply]
Thanks so much for the feedback!  To make sure I understand, so the reason this isn't implemented for file based provisioned alert rules as opposed to dashboard is a technical reason that the state of the provisioned alerts is constantly being refreshed from the file, so if the alert rule was in the process of being edited it would lose all the changes made from the UI.  Is that correct?

If this is the case then it seems that the workaround I described might be a permanent alternative.  In this case, it should be updated to reflect the most current version of the API.  To achieve that, could you point me to any doc that might explain how to grant the permissions needed for the /api/v1/provisioning [endpoint](https://grafana.com/docs/grafana/latest/developers/http_api/alerting_provisioning/) so as to avoid errors similar to `{"accessErrorId":"ACE0032154373","message":"You'll need additional permissions to perform this action. Permissions needed: alert.provisioning:read","title":"Access denied"}` and be able to do this the right way?  Thanks!

**Reply by sarasensible** [reply]
To address @alexweav 's point, we actually worked on this problem explicitly at my organization and implemented a nightly snapshot of the sqlite database that would be reloaded on startup to bring back any user-made changes.  We have since switched to an HA setup with a proper RDS database with regular snapshots, so we do not anticipate having this reloading issue going forward.

**Reply by sarasensible** [reply]
While I very much appreciate the desire to keep users from running into technical obstacles, the reason we will continue to use the workaround I provided instead of using the file based provisioning you describe is due to having somewhat opposing needs - 1) to have the alert configurations available from the start with an automatically provisioned Grafana and 2) once Grafana is up and running, to be able to iterate on the alert configuration and only periodically export this back to the file.  

We are presuming that not everyone who is on call and needs to configure the alert is sophisticated enough with the Terraform file templates to be able to make edits directly to the file based alert configuration, which creates an additional burden for the manager of those templates to have to intervene in every situation where an update is needed.

I'm not asserting that our user story preempts the valid reasons you both described, but I wanted to record our perspective so that the tradeoffs of the decision that is being made are clear.

**Reply by JohnnyQQQQ** [reply]
This is not specific to alert rules, it's any alerting object (alert rules, contact points, text templates, mute timings, notification policies), datasources, etc. 

A lot of users are running Grafana in a stateless way using things like Kubernetes and config maps mounted to the filesystem to provision objects, so we expect Grafana to be ephemeral. On each restart, Grafana would read the current provisioning files from the filesystem and overwrite any change. We have hundreds of thousands of active Grafana installations, so we have a very conservative stance regarding this.

As @alexweav mentioned, if we would implement Dashboards again today, we probably wouldn't make them editable when provisioned.

Regarding the error you are facing, you can check the scopes here https://grafana.com/docs/grafana/latest/administration/roles-and-permissions/access-control/custom-role-actions-scopes/

**Reply by sarasensible** [reply]
Understood, how about just exporting to a file from an alert rule created in the GUI?

I am using an Administrator Grafana token to call the new API, so I would think I have all the permissions I need.  However we are using the OSS version, is that API an Enterprise only feature?  If it is included in the OSS version, I can open a separate issue to try to track that error down.

**Reply by JohnnyQQQQ** [reply]
If you are facing problems with the API token feel free to open an issue reporting the bug with a detailed explanation of which token you used, how you created it, and which route you tried to access, then we can take a look.

**Reply by sarasensible** [reply]
Thanks for the recommendation, I am not able to reproduce this now so I am good to go.  Thanks so much for the help!

Update:  I figured out what was happening with the API - if the alert rule folder does not exist already, the API gives a misleading error:
```
{"accessErrorId":"ACE0771207682","message":"You'll need additional permissions to perform this action. Permissions needed: any of alert.rules:write, alert.rules:create, alert.rules:delete","title":"Access denied"}
```
So the alert rules folder must exist before any of this solution can be applied.

**Reply by schuenckferreira** [reply]
Hi, I am facing some problems with alert provisioning.

I exported an existing alert to json and did the Tweaks informed by @sarasensible (thanks for that) and managed to provision the alert.

But now two things are happening, the first problem I already identified and did a workaround.

In my original alert I have a regex, the alert was not working after provisioning the new json, for some reason Grafana was ignoring two backslashes, I had to add in my json 8 backslashes instead of 4 for it to understand this as 2.

Original: 
percentageField!~"^([9][0-9]|[9][0-9]\\\\\\\\.?[0-9]*|[1-9][0-9][0-9]\\\\\\\\.?[0-9]*)$",

New json: 
percentageField!~"^([9][0-9]|[9][0-9]\\\\\\\\\\\\\\\\.?[0-9]*|[1-9][0-9][0-9]\\\\\\\\\\\\\\\\.?[0-9]*)$",

The second problem:
In my json I have the field: "no_data_state": “OK”

For some reason after provisioning the file Grafana just ignores this json field and when there is no data it is setting the alert with the "No data" status instead of “OK”.

Any idea why this is happening?

**Reply by sarasensible** [reply]
@schuenckferreira not sure why that would be happening, if you provision the alert manually does it work?  Are you getting data at all for that metric?  If the issue persists I would create a new issue for it and paste in the full JSON for the alert so that others can reproduce it.

**Reply by schuenckferreira** [reply]
Hi @sarasensible, sorry for the delay in replying.

In the image below I created the same alert, one manually and the other provisioning the json in the folder /etc/grafana/provisioning/alerting.

The configuration is the same for both and it should treat NoData as OK, but the provisioned alert is firing alert with a DatasourceNoData message.
<img width="1362" alt="alerts" src="https://user-images.githubusercontent.com/100692543/191713460-2faa5ea4-2419-4899-a1d4-044876773fb3.png">

If I call /api/ruler/grafana/api/v1/rules/ for the provisioned rule I can see that it was provisioned as "no_data_state": "NoData":
![image](https://user-images.githubusercontent.com/100692543/191714172-4a4cc672-1ded-46ee-a4ed-ff71fe9cd56c.png)

But looking at the json file that was placed in the folder /etc/grafana/provisioning/alerting we can see that in the file it was set "no_data_state": "OK":
![image](https://user-images.githubusercontent.com/100692543/191715002-cae88f46-3f50-47ca-bbd6-53bc82c5fa88.png)

It seems that grafana is simply ignoring and overriding the parameter that was set in the file.

**Reply by schuenckferreira** [reply]
Below is the json that was provisioned in the folder: 

{
  "apiVersion": 1,
  "groups": [
    {
      "name": "search-test",
      "folder": "test",
      "interval": "1m",
      "rules": [
        {
          "expr": "",
          "for": "48h",
          "labels": {
            "search_alert": "true"
          },
          "title": "Provisioned Test",
          "condition": "B",
          "data": [
            {
              "refId": "A",
              "queryType": "",
              "relativeTimeRange": {
                "from": 300,
                "to": 0
              },
              "datasourceUid": "P1809F7CD0C75ACF3",
              "model": {
                "expr": "group by (channel, field, percentageField, executed_at)(\n    hk_solr_index_status_prod_seconds_max{\n        timerange=\"today\", \n        percentageField!~\"^([9][0-9]|[9][0-9]\\\\.?[0-9]*|[1-9][0-9][0-9]\\\\.?[0-9]*)$\",\n        channel=~\"channel1-v16|channel2-v16\",\n        field!~\"updated_win|updated_offer|updated_pro|updated_s|updated_r\"\n    }\n)",
                "hide": false,
                "intervalMs": 1000,
                "maxDataPoints": 43200,
                "refId": "A"
              }
            },
            {
              "refId": "B",
              "queryType": "",
              "relativeTimeRange": {
                "from": 0,
                "to": 0
              },
              "datasourceUid": "-100",
              "model": {
                "conditions": [
                  {
                    "evaluator": {
                      "params": [
                        0
                      ],
                      "type": "gt"
                    },
                    "operator": {
                      "type": "and"
                    },
                    "query": {
                      "params": [
                        "A"
                      ]
                    },
                    "reducer": {
                      "params": [],
                      "type": "last"
                    },
                    "type": "query"
                  }
                ],
                "datasource": {
                  "type": "__expr__",
                  "uid": "-100"
                },
                "hide": false,
                "intervalMs": 1000,
                "maxDataPoints": 43200,
                "refId": "B",
                "type": "classic_conditions"
              }
            }
          ],
          "updated": "2022-09-15T16:23:03Z",
          "intervalSeconds": 60,
          "version": 1,
          "uid": "UmNByW74kz",
          "namespace_uid": "KXSblZiVk",
          "namespace_id": 188,
          "rule_group": "search-test",
          "no_data_state": "OK",
          "exec_err_state": "Alerting"
        }
      ]
    }
  ]
}

**Reply by JohnnyQQQQ** [reply]
Your casing is not correct.

https://github.com/grafana/grafana/blob/55ef34bee492368e389868d7daddb675c699c2b2/pkg/services/provisioning/alerting/rules_types.go#L76-L77

**Reply by schuenckferreira** [reply]
@JohnnyQQQQ, 
I replaced the names and it worked like a charm. Thanks!
from:
"no_data_state": "OK"
"exec_err_state": "Alerting" 
to:
"noDataState": "OK"
"execErrState": "Alerting"

**Comment by jgournet** [comment]
Reading this Feature Request: this is **exactly** what we would need too.
We currently provision our alerts and dashboards via Terraform -> K8s config maps. If we need to tweak an alert: we can't do it via the UI, so we have to painfully update json files, apply and pray that we got it right.

In previous grafana versions, we could just edit the dashboard (and its alerts), and click "save as JSON". An ominous message would appear saying "hey, this is a provisioned dashboard: any changes you do now will be lost".
This message is clear enough that users (note: only "authenticated users" can edit - so if we feel that our users are too dumb to read this message, we can just remove the edit rights from there) should know what it means. You can tweak the message a bit by adding "edited alerts will NOT be evaluated at all !" to make it even clearer.

=> could you guys consider re-adding this feature ? I feel that it would help greatly on the adoption of grafana alert rules : I'm in the process of migrating to it, and I can confidently say that it's a HUGE undertaking. Having that feature would make it a bit less painful

Thank you 

**edit**: Just discovered that clicking on "Copy" does the trick: just copy the alert rule into a new one, and then extract it via the export API, then re-upload just the right json parts... it's not "perfect", but good enough :)

**Comment by holmesb** [comment]
"_on each restart the state from the disk would be replayed -- if the new changes made in the UI are not there yet, the user will lose all their changes._" - @JohnnyQQQQ 

But the helm chart uses sidecars, so restarts are unlikely. In other words, configmaps containing dashboards & alerts are reloaded on-the-fly without restarts. The benefits make this small risk worthwhile: intuitive alert development using the UI - with gitops guardrails. This ability to edit provisioned **dashboards** has worked well for years. "_In the past, this has caused a lot of trouble_" - has it? We've never had an instance of a user losing dashboard work due to Grafana restarting. We only ever restart Grafana for upgrades, which we schedule off-peak with pre-warned maintenance.

"_we want...a single source of truth_" - I think you misunderstand what is being asked for. There would still be a single source of truth: git. We are not asking for the ability to persist\save changes using the UI, just edit settings, preview the effect, and [observe](https://github.com/grafana/grafana/discussions/66541) the declared state (yaml). Again, this already works well for dashboards.

@jgournet thanks for your workaround ("_copy the alert rule into a new one, and then extract it via the export API_"), but this is hardly intuitive for users. I assume they also have to delete this temporary one when provisioning.

**Comment by jgournet** [comment]
turns out the "copy alert" workaround does not even work when all alerts are provisioned, somehow ?
the button just is not there "sometimes" (I have yet to find out when it's there/not there ?)

Anyway, please consider those changes, that would really help with the adoption of grafana 9 (and 10, since old alerting will be fully deprecated by then)

**Reply by jgournet** [reply]
update:
the "copy alert" button disappears as soon as the alerts sidecar injects alerts.
It would be nice to prevent that :)

**Comment by holmesb** [comment]
"_old alerting will be fully deprecated_" - it will be removed completely. We've reverted to legacy alerting because unified alerting isn't gitops-friendly yet due to this and also [#66541](https://github.com/grafana/grafana/discussions/66541). I expect there'll be increasing numbers of users stranded on 9.x until these features are added.

**Comment by danfinn** [comment]
We would like this feature as well.  We currently use a similar workflow with dashboards, they are provisioned from a file in git however if we need to we can make tweaks to them via the UI and then save the JSON and commit that to git.

The ability to do the same thing with alerts would be very handy.  Right now it's a bit clunky if you want to edit or tweak an existing alert that has been provisioned from a file.  You would have to copy it and then make your changes, then export the yaml and manually remove the old alert from the yaml so that it's not duplicated.

**Reply by rmawhinnie** [reply]
Love the idea :D watching

**Comment by holmesb** [comment]
The world is moving towards GitOps, yet with v10's unified alerts, Grafana has moved away from it. Pre v10, could provision gitops dashboards and alerts (legacy), and still change them intuitively using the UI. You couldn't save them because the single source of truth remained git. Instead users were shown the code-change their UI alteration would cause, and they could flow that through their git/deployment pipeline, including capture of rationale, peer review (via PR), approval gates, easy rollback, collaboration via code-sharing was easier, Grafana was stateless and easy to manage, etc, etc. Now all these benefits only really apply to dashboards. v10 discourages gitops for alerts because they aren't as intuitive as legacy ones. You have to always consult the API\provisioning documentation before making changes.

I think there's a fundamental misunderstanding in @JohnnyQQQQ's answer. We aren't asking for two sources of truth. Git would remain the single source of truth for provisioned alerts. We are only asking for the ability to intuitively explore available alert options using the UI, eg drop down menus, text boxes, etc, then pressing Save\OK will not persist to disk, just [show ](https://github.com/grafana/grafana/discussions/66541)the yaml ready to be placed into git.

Any chance this can be moved into an issue and added to the roadmap? Or at least status changed to unanswered, as this discussion has legs. People will move to a competitor that is gitops-friendly, or stay on <=v9.x.

**Reply by sarasensible** [reply]
I agree with you 100%, while from Grafana's perspective this issue is "answered" there remains some further discussion around it.  I'll de-answer it :)

**Comment by armandgrillet** [comment]
Hey everyone,

The alerting team has discussed about what to do following your feedback. We agree that we should propose a way to edit a provisioned alert from the UI (without being able to save it) and see its JSON / HCL code to export it to a file and store it a Git repository. We will work on this feature over the next months, our goal is to have an improved import/export experience by Grafana 10.2 including this feature.

**Reply by sarasensible** [reply]
This is FANTASTIC news!  Thanks so much to the Grafana team for grokking this use case and to those on this thread who enthusiastically supported this feature request!!

**Reply by armandgrillet** [reply]
We are trying to get https://github.com/grafana/grafana/pull/71983 in Grafana 10.1, this should improve things by a lot.

**Reply by armandgrillet** [reply]
We are aware that https://github.com/grafana/grafana/pull/71983 is not the end of the story and we need to allow the flow `Editing a provisioned alert rule` => `Exporting the yaml of the updated alert rule`. We will keep on working on a better UX for 10.2 as stated 2 weeks ago.

**Comment by jgournet** [comment]
Half-related, but thought it might interest users of provisioned dashboards with alerts:
https://github.com/grafana/grafana/discussions/72178

that's both a feature request, and some kind of way to migrate legacy alerts.

**Comment by armandgrillet** [comment]
This is available in Grafana 10.2 (which will be released in the coming weeks), the feature is called "Modify export". This allows you to modify a provisioned alert rule and easily export the updated rule group as JSON, YAML, or HCL code.
![SS 2023-10-16 at 11 04 30](https://github.com/grafana/grafana/assets/2117580/18093e80-bd0f-4439-80b4-796f4df36cf4)
![SS 2023-10-16 at 11 05 31](https://github.com/grafana/grafana/assets/2117580/f2ad1560-bb06-467b-bb2f-bc9851dca4bd)

**Comment by holmesb** [comment]
Progress, but if an alert is provisioned, then better to align with provisioned dashboards: a Save button. We'd like the gitops experience to deviate as little as possible from users' expectation, which is to change an alert, and hit save. A save button is intuitive, whereas pressing More --> Modify Export requires instruction\explanation. Either way, the resulting dialogue with format selector looks good, perhaps can also make the wording consistent with provisioned dashboards' Save button:

![image](https://github.com/grafana/grafana/assets/5072156/0b35c9cc-8c50-4678-8499-b32a9fbe6b6a)

With this description: "_update your dashboard in corresponding provisioning source_", our users always knew what they needed to do (obvs swap instances of the word "dashboard" for "alert").

**Comment by danfinn** [comment]
Agreed. It would be ideal if the workflow and wording was the same between dashboards and alerting.
